packages/hardhat-core/src/builtin-tasks/compile.ts

```diff
@@ -161,9 +161,22 @@ subtask(TASK_COMPILE_SOLIDITY_READ_FILE)
   .addParam("absolutePath", undefined, undefined, types.string)
   .setAction(
     async ({ absolutePath }: { absolutePath: string }): Promise<string> => {
-      return fsExtra.readFile(absolutePath, {
-        encoding: "utf8",
-      });
+      try {
+        return await fsExtra.readFile(absolutePath, {
+          encoding: "utf8",
+        });
+      } catch (e) {
+        if (!fsExtra.existsSync(absolutePath)) {
+          throw new HardhatError(ERRORS.RESOLVER.FILE_NOT_FOUND);
+        }
+
+        if (fsExtra.lstatSync(absolutePath).isDirectory()) {
```

Comment:
We should use the async version of this method. In general we always use async functions for file-system stuff except in certain cases where the function _has_ to be sync.

Reply:
As mentioned in our chat I'd go sync here because;
- it's an error so the process will stop anyway, no reason to go async
- sync is more readable